plutus-core/plutus-core/src/PlutusCore.hs

```diff
@@ -148,7 +150,13 @@ applyProgram
     :: Semigroup a
     => Program tyname name uni fun a
     -> Program tyname name uni fun a
-    -> Maybe (Program tyname name uni fun a)
+    -> Either ApplyProgramError (Program tyname name uni fun a)
 applyProgram (Program a1 v1 t1) (Program a2 v2 t2) | v1 == v2
-  = Just $ Program (a1 <> a2) v1 (Apply (termAnn t1 <> termAnn t2) t1 t2)
-applyProgram _ _ = Nothing
+  = Right $ Program (a1 <> a2) v1 (Apply (termAnn t1 <> termAnn t2) t1 t2)
+applyProgram (Program _a1 v1 _t1) (Program _a2 v2 _t2) =
+    Left $ MkApplyProgramError v1 v2
+
+getAppliedProgram :: Either ApplyProgramError prog -> prog
```

Comment:
Not a huge fan of this, I think people can just deconstruct the either themselves how they want to. This is basically `either (error . show) id`

Reply:
I guess we do use it a bunch. I guess I'd rather have the more generalized
```
unsafeFromRight :: (Show e) => Either e a -> a
unsafeFromRight (Right a) = a
unsafeFromRight (Left e) = error $ show e
```
or something

Reply:
If it's so general can I put it in the prelude instead?